File: ci/experiments/2026-09-20-0155/flutter/lib/local.dart
import 'dart:js_interop';

import 'package:genui/genui.dart';

import 'model_client.dart';
import 'models.dart';

// The two calls `web/webllm_bridge.js` puts on `window`. Everything crosses as
// a string, so there is no object graph to convert.
@JS('a2uiWebllm.create')
external JSPromise<JSAny?> _create(String modelId, JSFunction onProgress);

@JS('a2uiWebllm.chat')
external JSPromise<JSString> _chat(
  String messagesJson,
  double temperature,
  int maxTokens,
);

/// Runs the model inside this browser tab through WebLLM.
///
/// Nothing leaves the machine and no API key is needed, which is what Jane
/// wants when her key is not to hand. The costs are a download of a gigabyte or
/// more on the first run, and a browser with WebGPU.
class WebllmClient implements ModelClient {
  WebllmClient({required this.choice, required this.onStatus});

  final ModelChoice choice;
  final StatusCallback onStatus;

  bool _loaded = false;

  @override
  String get label => '${choice.modelId} (in this browser)';

  @override
  Future<String> send(String system, List<ChatMessage> history) async {
    if (!_loaded) {
      onStatus('Loading the model into this browser. The first run downloads '
          'it, which takes a while.');
      await _create(
        choice.modelId,
        ((JSString text) => onStatus(text.toDart)).toJS,
      ).toDart;
      _loaded = true;
    }

    onStatus('Thinking...');
    final reply = await _chat(
      webllmMessagesJson(system, history),
      choice.temperature,
      choice.maxOutputTokens,
    ).toDart;

    final text = reply.toDart;
    if (text.isEmpty) {
      throw StateError('The in-browser model returned nothing.');
    }
    return text;
  }
}

File: ci/experiments/2026-09-20-0155/react/src/local.ts
import type {LlmClient, Turn} from './gemini';
import type {ModelChoice} from './models';

/**
 * The local family, for when Jane does not have her key nearby.
 *
 * WebLLM downloads the weights into the browser on first use, which takes
 * minutes, so the engine is created lazily and the progress is reported to the
 * chat rather than hidden.
 */
/**
 * Fails early and clearly when this browser cannot run a model.
 *
 * Without it the failure arrives from inside WebLLM, much later and much less
 * clearly. The two halves fail differently: `navigator.gpu` is absent in an
 * older browser and on a page that is not a secure context, while a machine
 * with no usable GPU has `navigator.gpu` and hands back a null adapter. A
 * headless container is the second kind.
 */
async function requireWebGpu(): Promise<void> {
  const gpu = (navigator as {gpu?: {requestAdapter(): Promise<unknown>}}).gpu;
  if (!gpu) {
    throw new Error(
      'This browser has no WebGPU (navigator.gpu is undefined), which the ' +
        'in-browser model needs. Chrome or Edge 113+ over https or localhost ' +
        'can run it; otherwise pick a Gemini model.',
    );
  }
  const adapter = await gpu.requestAdapter().catch(() => null);
  if (!adapter) {
    throw new Error(
      'WebGPU is present but no GPU adapter is available, so the model cannot ' +
        'run in this browser. This is what a machine with no GPU, or a ' +
        'headless one, reports. Pick a Gemini model instead.',
    );
  }
}

export function webllmClient(modelId: string, onStatus: (s: string) => void): LlmClient {
  let engine: unknown;

  return {
    label: `${modelId} (in this browser)`,
    async send(system: string, turns: Turn[], choice: ModelChoice): Promise<string> {
      const webllm = await import('@mlc-ai/web-llm');
      if (!engine) {
        await requireWebGpu();
        onStatus('Downloading the model into this browser, this takes a few minutes...');
        engine = await webllm.CreateMLCEngine(modelId, {
          initProgressCallback: p => onStatus(p.text),
        });
      }
      onStatus('Thinking...');
      const reply = await (engine as {chat: any}).chat.completions.create({
        messages: [
          {role: 'system', content: system},
          ...turns.map(t => ({role: t.role === 'model' ? 'assistant' : 'user', content: t.text})),
        ],
        temperature: choice.temperature,
        max_tokens: choice.maxOutputTokens,
      });
      return reply.choices[0]?.message?.content ?? '';
    },
  };
}
